zebra-chain/src/keys/sapling.rs

```diff
@@ -153,7 +155,7 @@ fn zcash_h() -> jubjub::ExtendedPoint {
 /// Used to derive a diversified base point from a diversifier value.
 ///
 /// https://zips.z.cash/protocol/protocol.pdf#concretediversifyhash
-fn diversify_hash(d: [u8; 11]) -> Option<jubjub::ExtendedPoint> {
+pub fn diversify_hash(d: [u8; 11]) -> Option<jubjub::ExtendedPoint> {
```

Comment:
Is this function ever called outside of the `Diversifier`'s `From` impl? I think it would be preferable to encapsulate as a behavior of the `Diversifier` type there rather than exposing functions on bare byte arrays.

Also, why does it return an `Option`? I see that `jubjub_group_hash` returns an option, because it does hash-and-pray, but in the `From` impls below the option is unwrapped.  I checked the spec and didn't see any notes about this; the `None` variant is denoted by `\bottom` and is handled "outside" of the type signatures of the functions in the spec, so I don't see anything that prevents a diversifier from failing.  But I don't see any reason why it should succeed, either.

What does librustzcash do here?

Reply:
I made this public because I need to use it to construct a `NoteCommitment`, but forgot to use it! 🙊

![image](https://user-images.githubusercontent.com/552961/89471172-74c7ed80-d74b-11ea-9b65-c4496bb25936.png)

Reply:
Aha, past-me did the right thing! I'm making a comment to remind myself.

![image](https://user-images.githubusercontent.com/552961/89471621-9ecddf80-d74c-11ea-910d-203dcbeab29a.png)
![image](https://user-images.githubusercontent.com/552961/89471637-ac836500-d74c-11ea-97e9-6ed27a6a69d9.png)

Reply:
> Also, why does it return an `Option`? I see that `jubjub_group_hash` returns an option, because it does hash-and-pray, but in the `From` impls below the option is unwrapped. I checked the spec and didn't see any notes about this; the `None` variant is denoted by `\bottom` and is handled "outside" of the type signatures of the functions in the spec, so I don't see anything that prevents a diversifier from failing. But I don't see any reason why it should succeed, either.
> 
> What does librustzcash do here?

I'll go look 🔍

Reply:
@dconnolly Right, but when you're constructing a `NoteCommitment` wouldn't you want to be using a `Diversifier`, not the bare function? That's what I was asking about.

Reply:
That `unwrap` looks suspicious; g_d = ⊥ can certainly occur, for invalid addresses.

Reply:
@hdevalence most of the relevant functions have been moved into commitments::sapling::pedersen_hashes.rs, until a top level 'hash_functions' or similar module is created in the zebra-chain crate.

Reply:
> @dconnolly Right, but when you're constructing a `NoteCommitment` wouldn't you want to be using a `Diversifier`, not the bare function? That's what I was asking about.

Yep, I made `diversify_hash()` private again and just do `AffinePoint::from(Diversifier)` which calls it under the hood.

Reply:
> Also, why does it return an `Option`? I see that `jubjub_group_hash` returns an option, because it does hash-and-pray, but in the `From` impls below the option is unwrapped. I checked the spec and didn't see any notes about this; the `None` variant is denoted by `\bottom` and is handled "outside" of the type signatures of the functions in the spec, so I don't see anything that prevents a diversifier from failing. But I don't see any reason why it should succeed, either.
> 
> What does librustzcash do here?

Looks like [they also return an `Option`](https://github.com/zcash/librustzcash/blob/b02cf3b467524fd40515f893d060b8f750d21a68/zcash_primitives/src/primitives.rs#L110):

Reply:
> That `unwrap` looks suspicious; g_d = ⊥ can certainly occur, for invalid addresses. GroupHash has a ~50% chance of returning ⊥ for a given input.

When generating a new `Diversifier` or deriving one from a `SpendingKey`, we loop forever until we get `Some(result)`:

https://github.com/ZcashFoundation/zebra/pull/596/files#diff-2086d84326cc6e5847ae53a86c7f8742R648-R703

Reply:
Linking [this change](https://github.com/ZcashFoundation/zebra/pull/596/commits/b2dd674c6f6ff9a54d2ebb8a9d3ddb8804bd500e) here too